### ISSUE TITLE
Improve doc over MaxConcurrentReconciles

### ISSUE THREAD

**Issue by camilamacedo86** [kubernetes-sigs/controller-runtime#1416]
Would be nice be very nice we are able to provide further information for MaxConcurrentReconciles [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controller.go#L37) for we have a better go doc [here](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller#Options).  

People have been asking when/how using this option. We might add that `It is not possible to have two or more reconcile loops that handle the same object at the same time` and some other further explanations. See that people have also been providing further docs about the topic and would be nice to have something official. e.g: https://openkruise.io/en-us/blog/blog2.html

**Comment by vincent-pli** [comment]
@camilamacedo86 
Any process of issue,  
i want to share my experience since it suffer me for several days: if you want to enhance the performance of your `controller`, you could set `MaxConcurrentReconciles` to a big number, but do not forget to increase the `QPS` and `Burst` who is for `restClient`. or you `reconcile` will block when they try to update resource at the end of reconciling.

**Comment by MadhavJivrajani** [comment]
> It is not possible to have two or more reconcile loops that handle the same object at the same time

Hey @camilamacedo86! Thanks for attaching that blog post, it was very helpful and informative. However from what I understand, it *is* possible to have two or more loops handling the same object due to the work queue implementation? And considering the `dirty` and `processing` structures, not sure if the same object can actually be handled by two separate loops at the same time. Please correct me if I've misunderstood something here. Thanks!

**Comment by kevinfrommelt** [comment]
The workqueue is specifically designed to protect against concurrently processing the same object. When an object is added to the workqueue, it first gets placed in the `dirty` set and gets enqueued only if the object is not in the `processing` set. When that object gets pulled from the queue, it is transitioned from the `dirty` set to the `processing` set. This prevents the object from being enqueued if it is added again. Once the object is finished processing, it is removed from `processing` and if it is in the `dirty` set, it will be added to the queue again.
Relevant source code: https://github.com/kubernetes/client-go/blob/v0.22.0/util/workqueue/queue.go#L113-L182

This comment also implies that the same object can't be processed at the same time https://github.com/kubernetes-sigs/controller-runtime/blob/v0.9.6/pkg/internal/controller/controller.go#L213

**Comment by FillZpp** [comment]
/remove-lifecycle stale

> People have been asking when/how using this option. We might add that It is not possible to have two or more reconcile loops that handle the same object at the same time and some other further explanations. See that people have also been providing further docs about the topic and would be nice to have something official. e.g: https://openkruise.io/en-us/blog/blog2.html

Also the blog link we have changed to https://openkruise.io/blog/learning-concurrent-reconciling now :)

**Comment by schrej** [comment]
I think this would still be valuable.
/reopen
/remove-lifecycle rotten
